resource-managers/yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala

```diff
@@ -161,6 +170,27 @@ private[spark] abstract class YarnSchedulerBackend(
     totalRegisteredExecutors.get() >= totalExpectedExecutors * minRegisteredRatio
   }
 
+  override def getMergerLocations(
+      numPartitions: Int,
+      resourceProfileId: Int): Seq[BlockManagerId] = {
+    // Currently this is naive way of calculating numMergersNeeded for a stage. In future,
+    // we can use better heuristics to calculate numMergersNeeded for a stage.
+    val tasksPerExecutor = sc.resourceProfileManager
+        .resourceProfileFromId(resourceProfileId).maxTasksPerExecutor(sc.conf)
+    val numMergersNeeded = math.min(
+      math.max(1, math.ceil(numPartitions / tasksPerExecutor).toInt), maxNumExecutors)
+    val minMergersThreshold = math.max(minMergersStaticThreshold,
+      math.floor(numMergersNeeded * minMergersThresholdRatio).toInt)
+    val mergerLocations = blockManagerMaster
+      .getMergerLocations(numMergersNeeded, scheduler.nodeBlacklist())
```

Comment:
Can we add a note on what the constraints for invoking this method being invoked ? (in SchedulerBackend.scala preferably).
For example: 
- Will it happen with stage retries ?
- Will it happen with shuffle retry (same shuffleid, different stage's - when computed in different jobs) ?
, etc.

This impacts some of the discussion we might have here.
For example, 
* As blacklist is dynamic, how does this interact with stage retries ?
* As `blockManagerMaster.getMergerLocations` is randomizing the merger locations, how does it impact the relative ordering of mergers ?
, etc.

Reply:
I have explained a bit on the `SchedulerBackend#getShufflePushMergerLocations`, please take a look if this looks good or needs to be further expanded.